Try extracting load-script to a new package

scruffian opened the pull request:
#### Changes proposed in this Pull Request

This extracts `load-script` to a new package, which will make it easier to reuse other modules as packages (specifically Analytics).

#### Testing instructions

* Check that sections of Calypso that use `load-script` still work:
- /log-in
- /checkout
- location search
- /stats
- embeds
- facebook button
- analytics
- directly
- translator jumpstart
- geocoding
- videopress preview

scruffian committed:
Try extracting load-script to a new package

scruffian committed:
Use the built packages for the published module

scruffian committed:
Trying to fix tests

scruffian committed:
Add a wrapper for loading jquery on desktop

jsnajdr committed:
Fixup tests that fail because they mock lib/load-script

jsnajdr commented:
@scruffian I pushed a commit that makes all tests green. It was a simple change that changes the name of the module to be mocked.

scruffian commented:
@jsnajdr thanks! What are your thoughts on the loadJqueryDependentScriptDesktopWrapper lib, and how to test it?

jsnajdr commented:
> What are your thoughts on the loadJqueryDependentScriptDesktopWrapper lib, and how to test it?

Do you also need it in Jetpack? If not, I'd rather keep it in Calypso.

The unit tests for `loadScript` are not great. They test that internal implementation functions (`createScriptElement`, `executeCallbacks`) have been called, rather than the actual behavior. I don't know how to improve them at this moment though.

scruffian commented:
> Do you also need it in Jetpack? If not, I'd rather keep it in Calypso.

I agree. I'll remove the tests for that function I think...

scruffian committed:
Remove redundant tests

dmsnell commented on client/lib/load-jquery-dependent-script-desktop-wrapper/index.js:
```diff
@@ -0,0 +1,31 @@
+		asyncRequire( 'jquery', $ => {
+			window.$ = window.jQuery = $;
+			loadScript( url, callback );
+		} );
```
where are we getting `asyncRequire()`?

Is it supposed to be a global?

dmsnell reviewed (commented):
When reviewing I had hoped that we could think about eliminating this module entirely but I believe it's being used explicitly to load external scripts, something that our beloved `import()` function cannot do.

That being said as we think about moving it into its own package with a stronger contract than it has previously had did you consider turning it into a `Promise`-based API?

I wouldn't encourage doing this all in one go: moving it to a package _and_ turning it into a `Promise`, but I worry that if moving it first then changing to a `Promise` would happen after something in Jetpack or some external dependency is already using the callback semantic.

What's your timeline and rush on this? It looks feasible that we could turn it into a `Promise`-based interface before moving it to a package and ensure that no external libraries use the callback form.

Alternatively we could leave the callback API in place as we did with `wpcom.js` and gradually move code, though that has its own tradeoffs (like never ever being done).

Why? A good selection of existing uses of `loadScript()` are manually turning it into a `Promise` or using it in a way to which the `Promise` naturally maps.

```js
loadjQueryDependentScriptDesktopWrapper( SLIDESHOW_URLS.CYCLE_JS, () => {
	createSlideshow();
} );

// becomes
loadjQuery()
	.then( loadScript( SLIDESHOW_URLS.CYCLE_JS ) )
	.then( createSlideshow );
```

```js
const loadScript = promisify( loadScriptCallback );
await loadScript( src );

// becomes
await loadScript( src );
```

```js
function loadDirectlyScript() {
	return new Promise( ( resolve, reject ) => {
		loadScript( DIRECTLY_RTM_SCRIPT_URL, function( error ) {
			if ( error ) {
				return reject( new Error( `Failed to load script "${ error.src }".` ) );
			}
			resolve();
		} );
	} );
}

…
return loadDirctlyScript();

// becomes
return loadScript( DIRECTLY_RTM_SCRIPT_URL );
```

and more…

thoughts?

scruffian commented:
> What's your timeline and rush on this?

I'm already a long way down a rabbit hole here. I am moving loadScript, to a package, so that I can move the analytics lib to a package, so that I can use it in Jetpack, so that I can get Tracks working on Gutenberg blocks. Your suggestion sounds like a good one, but my concern is that I'm already so far away from my original aim that it's hard to justify spending more time on it!

dmsnell commented:
> I'm already a long way down a rabbit hole here. I am moving loadScript, to a package, so that I can move the analytics lib to a package, so that I can use it in Jetpack, so that I can get Tracks working on Gutenberg blocks. Your suggestion sounds like a good one, but my concern is that I'm already so far away from my original aim that it's hard to justify spending more time on it!

Sounds like it. I would probably recommend recreating `loadScript` in Jetpack then instead of trying to get this changed if you don't have the time to continue. That doesn't seem like a bad compromise to me either. In Jetpack you could create it almost by copying and pasting the code and you could start out with a `Promise` as I'm sure the calling code will want to have it and then you won't have to worry about legacy code.

Most of my objection to shipping this as-is has to do with the additional weight placed on the API and contract for `loadScript` when it becomes a consumed package. That is, _if_ we're going to lock it in place then now seems like the best time to update it and to update it internally _before_ packing; if that's too big of a task for now then let's just let it be…

ockham commented:
> > I'm already a long way down a rabbit hole here. I am moving loadScript, to a package, so that I can move the analytics lib to a package, so that I can use it in Jetpack, so that I can get Tracks working on Gutenberg blocks. Your suggestion sounds like a good one, but my concern is that I'm already so far away from my original aim that it's hard to justify spending more time on it!
> 
> Sounds like it. I would probably recommend recreating `loadScript` in Jetpack then instead of trying to get this changed if you don't have the time to continue. That doesn't seem like a bad compromise to me either. In Jetpack you could create it almost by copying and pasting the code and you could start out with a `Promise` as I'm sure the calling code will want to have it and then you won't have to worry about legacy code.

I'll voice a dissenting opinion here :sweat_smile: IMO it's preferable to package this lib from within Calypso and reuse it from Jetpack, even if its API might not yet be in the most desirable shape yet. This at least establishes the connection between Calypso and this new package as the source of truth, and Jetpack as the consumer. It also sets a path for future changes: If we want to change the API, we can do so, and release a new (major) version. Jetpack can then upgrade to the new package at will (and update the calls to the lib accordingly).

OTOH, if we duplicate this lib in Jetpack, we do just that: duplicate it. Nowhere is it then documented that that code is doing the same as the library we already have in Calypso. React components code in Jetpack is a perfect example of this: It's been copied over from Calypso (for a while living in the `dops-components` repo) and has been lying there for years now, getting out of sync with the maintained components in Calypso.

Note that for the present use case, the above is actually somewhat academic: @scruffian wants to use `analytics` from Jetpack, which requires `load-script` -- so Jetpack would use the latter lib only indirectly. So if we were to re-implement a (promise-based) `load-script` in Jetpack, we'd then either need to also re-implement `load-script` there (yet more duplication!), or adapt our prospective Calypso-originated `analytics` npm to use a different `load-script` implementation, and that's either going to mean we add an interface and logic just to allow that, or use some weird hacky way (webpack aliases or whatnot). None of this sounds great IMO.

> Most of my objection to shipping this as-is has to do with the additional weight placed on the API and contract for `loadScript` when it becomes a consumed package. That is, _if_ we're going to lock it in place then now seems like the best time to update it and to update it internally _before_ packing; if that's too big of a task for now then let's just let it be…

I used to be in that camp, but I've come to adopt a different stance since we're almost always finding ourselves in a situation as @scruffian: we have a concrete need for a Calypso-originated lib elsewhere, often pretty far down a dependency rabbit hole already. This presents us with an opportunity to package and reuse something, thus establishing and documenting aforementioned connections between libraries and consumers. OTOH, it's certainly not a position where we want to block ourselves by spending yet more time and thought on changing a lib's interface or implementation.

I was in fact tempted to do the same in few cases! Jetpack needs a `ProgressBar` (#33154) -- hmm, but shouldn't we change its implementation (from a `div` and a couple `aria` attributes) to an HTML5 `<progress />` element first? I still _want_ to do that eventually, but IMO, given the choice between blocking code re-use, or coming back later to release a new version with an improved interface and/or implementation, the later is preferable.

jsnajdr committed:
Simplify the executeCallbacks method

- `error` is a more fitting parameter name than `callbackArguments`.
- iterating over a `Set` can be more effective.

jsnajdr committed:
Make the code of onload/onerror handlers more consistent

The target element is both bound to `this` and received as `event.target`.
Use `this` consistently and don't mix both approaches within one function.

In the unit test, pass the more realistic element object as `this`.

jsnajdr committed:
Promisify the API of loadScript and loadjQueryDependentScript

Use the convention already used by the wpcom.js library: callback is the last and
optional argument. If it's supplied, invoke it and return `undefined`. If it's not
supplied, assume the promisifed version and return a `Promise`.

jsnajdr commented:
I'm pushing a few commits that implement the wpcom.js convention for promisified API: if the callback is not supplied (as the last and optional argument), then return a Promise. If it is supplied, call it (using the Node convention with `error` as first param) and return nothing (i.e., `undefined`).

The `loadjQueryDependentScript` test was a bit tricky to fix: the Promise resolution timing requires some magical `setTimeout( 0 )` calls.

There are opportunities to improve the library internally. For example, the map that tracks the registered callbacks? `Promise.then` has all that logic built in.

Also, I think we have a bug where the script will be loaded again when `loadScript( url )` is called after the first load is already finished.

But these can be postponed, as the API is stabilized and is not going to change.

jsnajdr committed:
Promisify loadScript calls in FB/Google social buttons

jsnajdr commented:
I tried to use the new promisified API for social login buttons for Google and FB (which is currently not used). Works well: the Google login button is disabled until the script loads:

<img width="408" alt="Screenshot 2019-05-22 at 11 56 37" src="https://user-images.githubusercontent.com/664258/58166430-191a5d00-7c8a-11e9-9973-5bf89ab7db94.png">

I also tested one place where `loadScript` remains callback-ish: location search in Classic Editor sidebar:

<img width="260" alt="Screenshot 2019-05-22 at 11 59 55" src="https://user-images.githubusercontent.com/664258/58166461-2b949680-7c8a-11e9-8ee4-99b3effdf6c5.png">

jsnajdr reviewed (approved):
I think this is ready to ship: the package works well and has a stable, backward-compatible and future-proof API.

As I have a little conflict of interest as a co-author, someone else should review and approve this, too. @ockham or @dmsnell?

ockham committed:
Turn SIFTSCIENCE_URL into a const

ockham committed:
Update README to fix import path

ockham committed:
Prettify README

ockham reviewed (approved):
Thanks! I think with your changes @jsnajdr, we have an API that should dispel @dmsnell's concerns :+1: 

This is looking good to me! I gave this a quick spin (testing [these use cases](https://github.com/Automattic/wp-calypso/pull/33155#issuecomment-494740131) specifically). This appears to be working well for me, and the code also looks good!

The remaining lint errors are unrelated. I did add 3 (minor) commits, irrelevant enough that I feel comfortable approving of this PR :heavy_check_mark: :slightly_smiling_face:
